### PR TITLE
jv-linked-list HW solution

### PR DESCRIPTION
**Pull request by luk14n**

### CODE REVIEW

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 1]
```diff
@@ -3,45 +3,144 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    static class Node<T> {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:10]: private
Private methods and classes should be after public ones in your class.

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 2]
```diff
@@ -3,45 +3,144 @@
...
+            tail = node.prev;
+            tail.next = null;
+        } else {
+            node.prev.next = node.next;
+            node.next.prev = node.prev;
+        }
+        size--;
+    }
+
+    private void checkIndex(int index) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:55]: Private methods and classes should be after public ones in your class.

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 3]
```diff
@@ -3,45 +3,144 @@
...
+        if ((index >= size) || (index < 0)) {
+            throw new IndexOutOfBoundsException("Wrong index");
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:57]: msg should be more informative

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 4]
```diff
@@ -3,45 +3,144 @@
...
+        for (T t : list) {
+            add(t);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNode(index).value;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:104]: This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the get method.

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Comment by luk14n** [comment]
can you pls also check my older HW from OOP?
 [https://github.com/mate-academy/jv-oop-advanced/pull/1251](url)
 Thanks beforehand

**Comment by luk14n** [comment]
can you pls also check my older HW from OOP?
[https://github.com/mate-academy/jv-oop-advanced/pull/1251](url)
Thanks beforehand

File: src/main/java/core/basesyntax/MyLinkedList.java (hunk 1 above)
Comment (Rommelua) [reply src/main/java/core/basesyntax/MyLinkedList.java:10]: not fixed

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 5]
```diff
@@ -3,45 +3,152 @@
...
+            add(value);
+            return;
+        }
+        checkIndex(index);
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            Node<T> currentNode = getNode(index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:47]: image

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 6]
```diff
@@ -3,45 +3,152 @@
...
     @Override
     public T get(int index) {
-        return null;
+        if ((index >= size) || (index < 0)) {
+            throw new IndexOutOfBoundsException("Index must be between 0 and "
+                    + (size - 1)
+                    + ", but was: "
+                    + index);
+        }
+        return getNode(index).value;
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:70]: image

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 7]
```diff
@@ -3,45 +3,152 @@
...
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = getNode(index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:76]: image

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 8]
```diff
@@ -3,45 +3,152 @@
...
+        T setValue = node.value;
+        node.value = value;
+        return setValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNode(index);
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:85]: image

File: src/main/java/core/basesyntax/MyLinkedList.java [hunk 9]
```diff
@@ -3,45 +3,152 @@
...
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == 0;
+    }
+
+    private Node<T> getNode(int index) {
```
Comment (Rommelua) [inline src/main/java/core/basesyntax/MyLinkedList.java:113]: call checkIndex here

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by Rommelua — APPROVED** [review]

**Pull request closed without merging**